src/main/java/gregtech/api/recipes/RecipeMap.java

```diff
@@ -167,19 +167,40 @@ public void addRecipe(ValidationResult<Recipe> validationResult) {
                 return;
         }
         Recipe recipe = validationResult.getResult();
-        recipeList.add(recipe);
-
-        for (FluidStack fluid : recipe.getFluidInputs()) {
-            recipeFluidMap.computeIfAbsent(new FluidKey(fluid), k -> new HashSet<>(1)).add(recipe);
+        if (recipeSet.add(recipe)) {
+            for (CountableIngredient countableIngredient : recipe.getInputs()) {
+                ItemStack[] stacks = countableIngredient.getIngredient().getMatchingStacks();
+                for (ItemStack itemStack : stacks) {
+                    ItemStackKey stackKey = KeySharedStack.getRegisteredStack(itemStack);
+                    recipeItemMap.computeIfPresent(stackKey, (k, v) -> {
+                        v.add(recipe);
+                        return v;
+                    });
+                    recipeItemMap.computeIfAbsent(stackKey, k -> new HashSet<>(1)).add(recipe);
```

Comment:
Is this explicit size parameter here (and one other place below) really necessary? Seems like it will pretty much always end up needing to resize immediately after, so isn't it better to just let Java handle the original size?